### ISSUE TITLE
helpers references parent component when nested not controller view_context

### ISSUE THREAD

**Issue by alhafoudh** [github/view_component#321]
### Steps to reproduce

parent_component.rb
```
class ParentComponent < ViewComponent::Base
end
```

parent_component.html.erb
```
<div id="parent"><%= helpers.current_user.name %></div>
<%= content %>
```

nested_component.rb
```
class NestedComponent < ViewComponent::Base
end
```

nested_component.html.erb
```
<div id="nested"><%= helpers.current_user.name %></div>
```

action.html.erb
```
<%= render ParentComponent.new do %>
  <%= render NestedComponent.new %>
<% end %>
```

### Expected behavior

I expect to see following output:

```
<div id="parent">John</div>
<div id="nested">John</div>
```

### Actual behavior

```
undefined local variable or method `current_user' for #<ParentComponent:0x00007f94f456c760>
```

### System configuration
**Rails version**: 6.0.2.2

**Ruby version**: 2.6.5p114

**Gem version**: 2.5.0

**Comment by joelhawksley** [comment]
@alhafoudh thank you for the report!

In this case, where would you expect `current_user` to be defined and thus be available to the component?

**Comment by kylefox** [comment]
@joelhawksley I put together a simple Rails app that demonstrates that `helpers` do not work properly in nested components: https://github.com/kylefox/nested-components-demo

This app defines a [`TableComponent`](https://github.com/kylefox/nested-components-demo/blob/master/app/components/table_component.html.erb), [`TableRowComponent`](https://github.com/kylefox/nested-components-demo/blob/master/app/components/table_row_component.html.erb), and a [`shout`](https://github.com/kylefox/nested-components-demo/blob/master/app/helpers/application_helper.rb#L2-L4) helper.

`TableComponent` accepts a collection and renders a `TableRowComponent` for each item. `TableRowComponent` uses the `shout` helper.

[It works](https://github.com/kylefox/nested-components-demo/blob/master/app/views/home/works.html.erb) when a `TableRowComponent` is rendered directly from the view:

```erb
<!-- home/works.html.erb -->
<table class="table">
  <thead>
    <th>First name</th>
    <th>Last name</th>
    <th>Email</th>
  </thead>
  <tbody>
    <% @people.each do |person| %>
      <%= render TableRowComponent.new(person: person) %>
    <% end %>
  </tbody>
</table>
```

But [this breaks](https://github.com/kylefox/nested-components-demo/blob/master/app/views/home/breaks.html.erb) (`undefined method `shout' for #<TableComponent:0x00007fb2459c3750>`), even though the template code for `TableComponent` is identical to `home/works.html.erb`:

```erb
<%= render TableComponent.new(people: @people) %>
```

Something funky definitely happens with `helpers`. If I output `<%= helpers.class %>` in the table row component, it outputs `#<Class:0x00007fb22587b750>` when being rendered directly in the view, but outputs `TableComponent` (the _parent component_ class) when being rendered within a parent component.

**Comment by joelhawksley** [comment]
@kylefox this unfortunately isn't a huge surprise. `helpers` is just an alias for `view_context`:

https://github.com/github/view_component/blob/6a4fb2083bef6efdfd6bf4e3a694a72e8c259f1d/lib/view_component/base.rb#L108-L111

In the nested case, the view context is the parent component, where global helpers are not available. 

This is ultimately a pretty significant bug. Do you have ideas for how we might go about solving it?

**Comment by franco** [comment]
I've redefined `helpers` in my project's base `ApplicationComponent` class:

```ruby
  # Restore access to global helpers
  def helpers
    view_context.controller.view_context
  end
```
I just needed a quick workaround to get going so I might doing something stupid here, but it solved the issue for me.

**Comment by joelhawksley** [comment]
@franco "escaping" to the controller view context is probably the right behavior here. Would you be willing to write a PR?

**Comment by franco** [comment]
@joelhawksley PR should fix it as discussed but be aware that this is potentially a breaking change. Some code might relay on the incorrect behaviour.

**Comment by woto** [comment]
Cool. Same issue. One thing. Pretty sure that this PR would have to change MAJOR version according https://semver.org/

**Comment by franco** [comment]
@woto I thought so too at first but [one could argue that it is just a malfunction from the intended behaviour](https://github.com/github/view_component/pull/374#issuecomment-644187953) and classify it simply as a bug. Anyways, the PR has been released already in [2.11.0](https://github.com/github/view_component/blob/master/CHANGELOG.md#2110).

**Comment by joelhawksley** [comment]
Closed by https://github.com/github/view_component/pull/374
